File: Corte 2/Ejercicios_Ordenamientos/ordenamiento.cpp
#include <iostream>
using namespace std;

int comparaciones = 0;
int intercambios = 0;


void resetContadores() {
    comparaciones = 0;
    intercambios = 0;
}

/* Bubble Sort */
void bubbleSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            comparaciones++;
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                intercambios++;
            }
        }
    }
}
/* Selection Sort */
void selectionSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            comparaciones++;
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        int temp = arr[minIdx];
        arr[minIdx] = arr[i];
        arr[i] = temp;
        intercambios++;
    }
}
/* Insertion Sort */
void insertionSort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            comparaciones++;
            arr[j + 1] = arr[j];
            intercambios++;
            j--;
        }
        arr[j + 1] = key;
        intercambios++;
    }
}

int main() {
    int datosOrig[] = {64, 25, 12, 22, 11, 90, 45, 33};
    int n = sizeof(datosOrig) / sizeof(datosOrig[0]);
    int arr[8];

    // Probar Bubble Sort
    for(int i=0; i<n; i++) arr[i] = datosOrig[i];
    resetContadores();
    bubbleSort(arr, n);
    cout << "Bubble Sort -> Comparaciones: " << comparaciones << ", Intercambios: " << intercambios << endl;

    // Probar Selection Sort
    for(int i=0; i<n; i++) arr[i] = datosOrig[i];
    resetContadores();
    selectionSort(arr, n);
    cout << "Selection Sort -> Comparaciones: " << comparaciones << ", Intercambios: " << intercambios << endl;

    // Probar Insertion Sort
    for(int i=0; i<n; i++) arr[i] = datosOrig[i];
    resetContadores();
    insertionSort(arr, n);
    cout << "Insertion Sort -> Comparaciones: " << comparaciones << ", Intercambios: " << intercambios << endl;

    return 0;
}

File: Corte 2/Ejercicios_Ordenamientos/ordenamiento.py
# comparo dos vecinos si estan al reves los intercambio y repito el proceso 
# hast que nadie mas se mueva bubble sort

comparaciones = 0
intercambios = 0

def bubble_sort(arr):
    global comparaciones, intercambios
    comparaciones = 0
    intercambios = 0
    n = len(arr)
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            comparaciones += 1
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                intercambios += 1
                swapped = True
        if not swapped:
            break
    return arr
#buscar el más pequeño y ponerlo al 
# principio selection sort

def selection_sort(arr):
    global comparaciones, intercambios
    comparaciones = 0
    intercambios = 0
    n = len(arr)
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            comparaciones += 1
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
        intercambios += 1
    return arr
# tomo el segundo y lo inserto donde va 
# respecto al primero insertion sort

def insertion_sort(arr):
    global comparaciones, intercambios
    comparaciones = 0
    intercambios = 0
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
            comparaciones += 1
        arr[j + 1] = key
        intercambios += 1
    return arr

# datos: [64, 25, 12, 22, 11, 90, 45, 33]

Lista = [64, 25, 12, 22, 11, 90, 45, 33]


print("Lista original:", Lista)
print("Lista ordenada (Bubble Sort):", bubble_sort(Lista.copy()))
print("Numero de comparaciones en bubble sort: ", comparaciones)
print("Numero de intercambios en bubble sort: ", intercambios)

print()

print("Lista ordenada (Selection Sort):", selection_sort(Lista.copy()))
print("Numero de comparaciones en selection sort: ", comparaciones)
print("Numero de intercambios en selection sort: ", intercambios)

print()

print("Lista ordenada (Insertion Sort):", insertion_sort(Lista.copy()))
print("Numero de comparaciones en insertion sort: ", comparaciones)
print("Numero de intercambios en insertion sort: ", intercambios)